File: examples/ai-documentation-consumer.js
#!/usr/bin/env node

/**
 * Example: AI Documentation Consumer
 *
 * This script demonstrates how AI agents or automated tools can consume
 * the documentation API to understand the application's architecture.
 *
 * Usage:
 *   # Make sure dev server is running
 *   npm run dev
 *
 *   # In another terminal
 *   node examples/ai-documentation-consumer.js
 */

const API_BASE_URL = 'http://localhost:3000'

/**
 * Fetches the complete API documentation in Markdown format
 */
async function fetchArchitectureDoc() {
  console.log('📚 Fetching API Architecture Documentation...\n')

  try {
    const response = await fetch(`${API_BASE_URL}/api/docs/architecture`)

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`)
    }

    const markdown = await response.text()

    console.log('✅ Documentation fetched successfully!')
    console.log(`📄 Size: ${(markdown.length / 1024).toFixed(2)} KB`)
    console.log(`📝 Lines: ${markdown.split('\n').length}\n`)

    // Example: Extract specific sections
    const sections = extractSections(markdown)
    console.log('📋 Documentation Sections:')
    sections.forEach(section => console.log(`   - ${section}`))

    return markdown
  } catch (error) {
    console.error('❌ Error fetching documentation:', error.message)
    throw error
  }
}

/**
 * Fetches dynamically generated documentation metadata
 */
async function fetchGeneratedDoc() {
  console.log('\n🔧 Fetching Generated Documentation...\n')

  try {
    const response = await fetch(`${API_BASE_URL}/api/docs/generate`)

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`)
    }

    const data = await response.json()

    console.log('✅ Generated documentation fetched successfully!')
    console.log(`🕒 Generated at: ${new Date(data.generatedAt).toLocaleString()}`)
    console.log(`📡 Endpoints: ${data.endpoints.length}`)
    console.log(`🔧 Utilities: ${data.utilities.length}`)
    console.log(`📦 Type Interfaces: ${data.types.interfaces.length}\n`)

    // Display endpoints
    console.log('📡 Available Endpoints:')
    data.endpoints.forEach(endpoint => {
      console.log(`   ${endpoint.method} ${endpoint.path}`)
      if (endpoint.queryParams.length > 0) {
        console.log(`      Params: ${endpoint.queryParams.join(', ')}`)
      }
    })

    // Display utilities
    console.log('\n🔧 Server Utilities:')
    data.utilities.forEach(util => {
      console.log(`   ${util.file}`)
      console.log(`      Functions: ${util.functions.join(', ')}`)
    })

    // Display types
    console.log('\n📦 Type Interfaces:')
    data.types.interfaces.forEach(iface => {
      console.log(`   ${iface.name}`)
      console.log(`      Fields: ${iface.fields.length}`)
    })

    return data
  } catch (error) {
    console.error('❌ Error fetching generated documentation:', error.message)
    throw error
  }
}

/**
 * Example: AI-style analysis of the API
 */
async function analyzeAPI() {
  console.log('\n🤖 AI Analysis of API Architecture...\n')

  const data = await fetchGeneratedDoc()

  // Analyze endpoint patterns
  const methods = {}
  data.endpoints.forEach(endpoint => {
    methods[endpoint.method] = (methods[endpoint.method] || 0) + 1
  })

  console.log('📊 API Analysis:')
  console.log('\n   HTTP Methods:')
  Object.entries(methods).forEach(([method, count]) => {
    console.log(`      ${method}: ${count} endpoint(s)`)
  })

  // Identify key architectural components
  console.log('\n   Key Components:')
  const components = [
    { name: 'Data Fetching', file: 'googleSheets.ts' },
    { name: 'Person Identification', file: 'personIdentifier.ts' },
    { name: 'Installment Processing', file: 'installmentProcessor.ts' },
    { name: 'Filtering', file: 'transactionFilters.ts' }
  ]

  components.forEach(component => {
    const util = data.utilities.find(u => u.file === component.file)
    if (util) {
      console.log(`      ✓ ${component.name}: ${util.functions.length} function(s)`)
    } else {
      console.log(`      ✗ ${component.name}: Not found`)
    }
  })

  // Check Transaction type structure
  console.log('\n   Transaction Type:')
  const transactionType = data.types.interfaces.find(i => i.name === 'Transaction')
  if (transactionType) {
    console.log(`      Fields: ${transactionType.fields.length}`)
    transactionType.fields.slice(0, 5).forEach(field => {
      console.log(`         - ${field.trim()}`)
    })
    if (transactionType.fields.length > 5) {
      console.log(`         ... and ${transactionType.fields.length - 5} more`)
    }
  }
}

/**
 * Example: Search documentation for specific information
 */
async function searchDocumentation(query) {
  console.log(`\n🔍 Searching documentation for: "${query}"\n`)

  const markdown = await fetchArchitectureDoc()
  const lines = markdown.split('\n')

  const results = lines
    .map((line, index) => ({ line, index }))
    .filter(({ line }) => line.toLowerCase().includes(query.toLowerCase()))
    .slice(0, 10)

  if (results.length === 0) {
    console.log('   No results found.')
  } else {
    console.log(`   Found ${results.length} occurrences:\n`)
    results.forEach(({ line, index }) => {
      console.log(`   Line ${index + 1}: ${line.trim()}`)
    })
  }
}

/**
 * Example: Compare documented vs. actual endpoints
 */
async function verifyDocumentation() {
  console.log('\n✅ Verifying Documentation Accuracy...\n')

  const markdown = await fetchArchitectureDoc()
  const generated = await fetchGeneratedDoc()

  // Check if all generated endpoints are documented
  console.log('   Checking endpoint coverage:')

  let allDocumented = true
  generated.endpoints.forEach(endpoint => {
    const isDocumented = markdown.includes(`${endpoint.method} ${endpoint.path}`)
    const status = isDocumented ? '✓' : '✗'
    console.log(`      ${status} ${endpoint.method} ${endpoint.path}`)

    if (!isDocumented) {
      allDocumented = false
    }
  })

  if (allDocumented) {
    console.log('\n   ✅ All endpoints are documented!')
  } else {
    console.log('\n   ⚠️  Some endpoints are missing from documentation')
  }
}

/**
 * Helper: Extract section headers from markdown
 */
function extractSections(markdown) {
  const lines = markdown.split('\n')
  return lines
    .filter(line => line.startsWith('## '))
    .map(line => line.replace('## ', '').trim())
}

/**
 * Main execution
 */
async function main() {
  console.log('🤖 AI Documentation Consumer Example\n')
  console.log('This script demonstrates how to consume the documentation API\n')
  console.log('='=.repeat(70))

  try {
    // Example 1: Fetch and display architecture documentation
    await fetchArchitectureDoc()

    // Example 2: Fetch and analyze generated documentation
    await fetchGeneratedDoc()

    // Example 3: AI-style analysis
    await analyzeAPI()

    // Example 4: Search for specific information
    await searchDocumentation('query parameters')

    // Example 5: Verify documentation accuracy
    await verifyDocumentation()

    console.log('\n' + '='.repeat(70))
    console.log('✅ Documentation consumer example completed successfully!')
    console.log('\n💡 Use this pattern to integrate with AI agents, CI/CD, or other tools.')

  } catch (error) {
    console.error('\n❌ Example failed:', error.message)
    console.log('\n💡 Make sure the dev server is running: npm run dev')
    process.exit(1)
  }
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main()
}

// Export for use as module
export {
  fetchArchitectureDoc,
  fetchGeneratedDoc,
  analyzeAPI,
  searchDocumentation,
  verifyDocumentation
}

File: examples/ai-documentation-consumer.py
#!/usr/bin/env python3
"""
Example: AI Documentation Consumer (Python)

This script demonstrates how AI agents or automated tools written in Python
can consume the documentation API to understand the application's architecture.

Requirements:
    pip install requests

Usage:
    # Make sure dev server is running
    npm run dev

    # In another terminal
    python3 examples/ai-documentation-consumer.py
"""

import requests
import json
from datetime import datetime
from typing import Dict, List, Any

API_BASE_URL = 'http://localhost:3000'


def fetch_architecture_doc() -> str:
    """Fetches the complete API documentation in Markdown format."""
    print('📚 Fetching API Architecture Documentation...\n')

    try:
        response = requests.get(f'{API_BASE_URL}/api/docs/architecture')
        response.raise_for_status()

        markdown = response.text

        print('✅ Documentation fetched successfully!')
        print(f'📄 Size: {len(markdown) / 1024:.2f} KB')
        print(f'📝 Lines: {len(markdown.splitlines())}\n')

        # Extract sections
        sections = extract_sections(markdown)
        print('📋 Documentation Sections:')
        for section in sections:
            print(f'   - {section}')

        return markdown

    except requests.exceptions.RequestException as e:
        print(f'❌ Error fetching documentation: {e}')
        raise


def fetch_generated_doc() -> Dict[str, Any]:
    """Fetches dynamically generated documentation metadata."""
    print('\n🔧 Fetching Generated Documentation...\n')

    try:
        response = requests.get(f'{API_BASE_URL}/api/docs/generate')
        response.raise_for_status()

        data = response.json()

        print('✅ Generated documentation fetched successfully!')
        print(f'🕒 Generated at: {datetime.fromisoformat(data["generatedAt"].replace("Z", "+00:00")).strftime("%Y-%m-%d %H:%M:%S")}')
        print(f'📡 Endpoints: {len(data["endpoints"])}')
        print(f'🔧 Utilities: {len(data["utilities"])}')
        print(f'📦 Type Interfaces: {len(data["types"]["interfaces"])}\n')

        # Display endpoints
        print('📡 Available Endpoints:')
        for endpoint in data['endpoints']:
            print(f'   {endpoint["method"]} {endpoint["path"]}')
            if endpoint.get('queryParams'):
                print(f'      Params: {", ".join(endpoint["queryParams"])}')

        # Display utilities
        print('\n🔧 Server Utilities:')
        for util in data['utilities']:
            print(f'   {util["file"]}')
            print(f'      Functions: {", ".join(util["functions"])}')

        # Display types
        print('\n📦 Type Interfaces:')
        for iface in data['types']['interfaces']:
            print(f'   {iface["name"]}')
            print(f'      Fields: {len(iface["fields"])}')

        return data

    except requests.exceptions.RequestException as e:
        print(f'❌ Error fetching generated documentation: {e}')
        raise


def analyze_api(data: Dict[str, Any]) -> None:
    """AI-style analysis of the API."""
    print('\n🤖 AI Analysis of API Architecture...\n')

    # Analyze endpoint patterns
    methods = {}
    for endpoint in data['endpoints']:
        method = endpoint['method']
        methods[method] = methods.get(method, 0) + 1

    print('📊 API Analysis:')
    print('\n   HTTP Methods:')
    for method, count in methods.items():
        print(f'      {method}: {count} endpoint(s)')

    # Identify key architectural components
    print('\n   Key Components:')
    components = [
        {'name': 'Data Fetching', 'file': 'googleSheets.ts'},
        {'name': 'Person Identification', 'file': 'personIdentifier.ts'},
        {'name': 'Installment Processing', 'file': 'installmentProcessor.ts'},
        {'name': 'Filtering', 'file': 'transactionFilters.ts'}
    ]

    for component in components:
        util = next((u for u in data['utilities'] if u['file'] == component['file']), None)
        if util:
            print(f'      ✓ {component["name"]}: {len(util["functions"])} function(s)')
        else:
            print(f'      ✗ {component["name"]}: Not found')

    # Check Transaction type structure
    print('\n   Transaction Type:')
    transaction_type = next(
        (i for i in data['types']['interfaces'] if i['name'] == 'Transaction'),
        None
    )

    if transaction_type:
        print(f'      Fields: {len(transaction_type["fields"])}')
        for field in transaction_type['fields'][:5]:
            print(f'         - {field.strip()}')
        if len(transaction_type['fields']) > 5:
            print(f'         ... and {len(transaction_type["fields"]) - 5} more')


def search_documentation(query: str, markdown: str) -> None:
    """Search documentation for specific information."""
    print(f'\n🔍 Searching documentation for: "{query}"\n')

    lines = markdown.splitlines()
    results = [
        {'line': line, 'index': i}
        for i, line in enumerate(lines)
        if query.lower() in line.lower()
    ][:10]

    if not results:
        print('   No results found.')
    else:
        print(f'   Found {len(results)} occurrences:\n')
        for result in results:
            print(f'   Line {result["index"] + 1}: {result["line"].strip()}')


def verify_documentation(markdown: str, data: Dict[str, Any]) -> None:
    """Compare documented vs. actual endpoints."""
    print('\n✅ Verifying Documentation Accuracy...\n')
    print('   Checking endpoint coverage:')

    all_documented = True

    for endpoint in data['endpoints']:
        endpoint_str = f'{endpoint["method"]} {endpoint["path"]}'
        is_documented = endpoint_str in markdown
        status = '✓' if is_documented else '✗'
        print(f'      {status} {endpoint_str}')

        if not is_documented:
            all_documented = False

    if all_documented:
        print('\n   ✅ All endpoints are documented!')
    else:
        print('\n   ⚠️  Some endpoints are missing from documentation')


def extract_sections(markdown: str) -> List[str]:
    """Extract section headers from markdown."""
    lines = markdown.splitlines()
    return [
        line.replace('## ', '').strip()
        for line in lines
        if line.startswith('## ')
    ]


def save_to_file(markdown: str, filename: str = 'api-documentation.md') -> None:
    """Save documentation to a file."""
    print(f'\n💾 Saving documentation to {filename}...')
    with open(filename, 'w', encoding='utf-8') as f:
        f.write(markdown)
    print(f'✅ Saved successfully!')


def export_metadata(data: Dict[str, Any], filename: str = 'api-metadata.json') -> None:
    """Export metadata to JSON file."""
    print(f'\n💾 Exporting metadata to {filename}...')
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
    print(f'✅ Exported successfully!')


def main():
    """Main execution."""
    print('🤖 AI Documentation Consumer Example (Python)\n')
    print('This script demonstrates how to consume the documentation API\n')
    print('=' * 70)

    try:
        # Example 1: Fetch and display architecture documentation
        markdown = fetch_architecture_doc()

        # Example 2: Fetch and analyze generated documentation
        generated = fetch_generated_doc()

        # Example 3: AI-style analysis
        analyze_api(generated)

        # Example 4: Search for specific information
        search_documentation('query parameters', markdown)

        # Example 5: Verify documentation accuracy
        verify_documentation(markdown, generated)

        # Optional: Save to files
        # save_to_file(markdown)
        # export_metadata(generated)

        print('\n' + '=' * 70)
        print('✅ Documentation consumer example completed successfully!')
        print('\n💡 Use this pattern to integrate with AI agents, CI/CD, or other tools.')

    except Exception as e:
        print(f'\n❌ Example failed: {e}')
        print('\n💡 Make sure the dev server is running: npm run dev')
        exit(1)


if __name__ == '__main__':
    main()
